### DISCUSSION TITLE
Custom Rendering?

### DISCUSSION THREAD

**Discussion by DasLixou** [bevyengine/bevy#6519] [Q&A] [answered]
So i want to render a tilemap. not like the crates bevy-ecs-tilemap etc. i have my own system i really wanna use. so my qwestion is what the smartest way of rendering tiles is. i have a tilemap with like lets just say 20 tiles down. should i make lets say 10_000_000 entites corresponding to tiles and then make then visible and invisible by if there is a solid tile above them? or should i make a chunk entity that holds tiles as structs (thats what i did in my cpp version) and then somehow render them manually (idk how to do that in bevy)? or is there a other way u would recommend to me? thanks :)

**Comment by zicklag** [comment, marked as answer]
I think chunking is useful, but mostly from a rendering standpoint, to avoid excessive draw calls.

You don't necessarily need to put your tiles in a chunk entity because, similar to `bevy-ecs-tilemap` you can have an individual entity per tile, and then convert it to chunks before rendering.

But it might be simpler to code if you just did an entity per chunk.

---

Anyway, you can absolutely do custom rendering in Bevy. I don't know of any good tutorials specific to Bevy, but Bevy uses [`wgpu`](https://wgpu.rs/) which you can find a few tutorials on. WGPU is a sane, safe rendering API, so if you have done a little bit of OpenGL rendering, you should be able to understand what the functions mean, and it shouldn't be too hard to get stuff rendering. It's _way_ nicer than OpenGL.

As for the Bevy specific stuff, I'd look at the source code for other Bevy plugins that do custom rendering. Maybe [`bevy_simple_tilemap`](https://github.com/forbjok/bevy_simple_tilemap).

**Reply by DasLixou** [reply]
Thanks! That’s a **perfect** answer. I’ll definitely read through wgpu stuff and your promise that it’s _way nicer_ than OpenGL sounds exciting to me :)

**Reply by DasLixou** [reply]
@zicklag is there a tutorial or book that helps with custom rendering things. bevy_simple_tilemap looks really hard to read to me.

**Reply by zicklag** [reply]
Bevy's made a lot of great changes ( and improvements! ) since I last got into it, so I'm a little bit out of the loop on specifics.

It looks like it actually might be quite a bit easier than it used to be. You used to have to create a custom pipeline, but I think you can get away without that now.

---

From what I can tell, the important things to look at are the [material example](https://github.com/bevyengine/bevy/blob/latest/examples/shader/shader_material.rs) and the [custom vertex attributes](https://github.com/bevyengine/bevy/blob/latest/examples/shader/custom_vertex_attribute.rs) example.

The material example will show you how to setup a custom shader which you can use for your tilemap chunks. It shows how to set textures and uniforms, which will allow you to get the information you need for your tileset size/grid/etc. That way you'll be able to setup the tile sampling in the shader.

You might also need the custom vertex attributes if you need to put any of your own tile data in with the vertices which you might need to if you can't fit all the info in the uniforms.

---

Also, if you need anything else check out the [Shaders section](https://github.com/bevyengine/bevy/tree/latest/examples#shaders) of the examples.

**Reply by DasLixou** [reply]
thanks. just amazing how fast and detailed and helpful your answers are <3

**Reply by DasLixou** [reply]
Isn't just a material stupid? like i probably need to have a custom render func that calls a shader 4 times (vertices) a tile and not a chunk, don't i?

**Reply by zicklag** [reply]
I've managed to render whole tilemap layers with only 4 vertices per chunk of tiles by being _slightly_ fancy with the UV sampling, If all the tiles in a chunk share the same tilemap texture.

For instance:

- Let's say your chunks are 3x3 tiles in size ( you'd have more in practice, but to match the simple diagram below ).
    - This means that you will have 9 tiles in each chunk.
- In a uniform, you can pass an array of 9 indexes, which are the index of the tile in your tilemap texture that is supposed to be rendered in the corresponding spot in your chunk.
- Then in your fragment shader, you calculate which portion of the tilemap texture you are supposed to sample from:
    - You use the UV of the fragment in the chunk to calculate the index of the tile in your chunk that your fragment is in
    - You use that index to find out which tilemap tile you are supposed to render
    - You use the UV of the fragment to figure out the offset of tilemap tile to sample

I'm sorry if that doesn't make any sense, because it's a little hard to explain. _Maybe_ this diagram will help:

https://excalidraw.com/#json=uSFRcmGZSrvZtVF4_hogi,2cuiHmX3UYW-1zi-5jboeQ

![image](https://user-images.githubusercontent.com/25393315/201424260-9573623c-6495-4ac5-b919-626fed97ca2c.png)

So it's perfectly possible to render a chunk of tiles, with only a shader ( and a plain, square mesh ).

---

At the same time, you could go the route of creating a mesh that uses 4 vertices per tile, and you put the tilemap tile index as a custom vertex attribute for each vertex.

Then in the fragment shader, instead of calculating where you are in the chunk and determining which tilemap index to use from an array, you can just read the custom vertex attribute.

That might be simpler, and I'm not expert at graphics performance, so I can't tell you for sure if it'd be slower, but it would have more vertices.
